dask/config.py

```diff
@@ -262,80 +262,104 @@ def ensure_file(source, destination=None, comment=True):
-            if value == "--delete--":
-                d = self.config
-                try:
-                    while len(keys) > 1:
-                        d = d[keys[0]]
-                        keys = keys[1:]
-                    del d[keys[0]]
-                except KeyError:
-                    pass
-            else:
-                self._assign(keys, value, self.config)
-
-    @classmethod
-    def _assign(cls, keys, value, d, old=None, path=[]):
-        """ Assign value into a nested configuration dictionary
+        for op, path, value in reversed(self._record):
+            d = self.config
+            if op == "replace":
+                for key in path[:-1]:
+                    d = d.setdefault(key, {})
+                d[path[-1]] = value
+            else:  # insert
+                for key in path[:-1]:
+                    try:
+                        d = d[key]
+                    except KeyError:
+                        break
+                else:
+                    d.pop(path[-1], None)
 
-        Optionally record the old values in old
+    def _assign(self, keys, value, d, path=(), record=True):
+        """Assign value into a nested configuration dictionary
 
         Parameters
         ----------
-        keys: Sequence[str]
-            The nested path of keys to assign the value, similar to toolz.put_in
-        value: object
-        d: dict
+        keys : Sequence[str]
+            The nested path of keys to assign the value.
+        value : object
+        d : dict
             The part of the nested dictionary into which we want to assign the
             value
-        old: dict, optional
-            If provided this will hold the old values
-        path: List[str]
-            Used internally to hold the path of old values
+        path : Tuple[str], optional
+            The path history up to this point.
+        record : bool, optional
+            Whether this operation needs to be recorded to allow for rollback.
         """
         key = canonical_name(keys[0], d)
+        path = path + (key,)
+
         if len(keys) == 1:
-            if old is not None:
-                path_key = tuple(path + [key])
+            if record:
                 if key in d:
-                    old[path_key] = d[key]
+                    self._record.append(("replace", path, d[key]))
                 else:
-                    old[path_key] = "--delete--"
+                    self._record.append(("insert", path, None))
             d[key] = value
         else:
             if key not in d:
+                if record:
+                    self._record.append(("insert", path, None))
                 d[key] = {}
-                if old is not None:
-                    old[tuple(path + [key])] = "--delete--"
-                old = None
-            cls._assign(keys[1:], value, d[key], path=path + [key], old=old)
+                # No need to record after an insert
+                record = False
```

Comment:
Wouldn't we want to remove it when we roll things back?  Clearly this works, because it's tested below.

Reply:
We record the insert at the level we insert it at, but don't record subsequent inserts down that path (since reversing the root insert is sufficient).

Reply:
I've pushed an updated comment that should hopefully clarify this.